main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala

```diff
@@ -0,0 +1,716 @@
+
+    /**
+      * Composes `this` equality result with `that` equality result.
+      *
+      * - Composes the substitution,
+      * - combines the leftover constraints, and
+      * - indicates progress if one of the two made progress.
+      */
+    def @@(that: ResolutionResult): ResolutionResult = {
+      val ResolutionResult(s1, cs1, p1) = this
+      val ResolutionResult(s2, cs2, p2) = that
+      ResolutionResult(s1 @@ s2, cs1 ++ cs2, p1 || p2)
+    }
+  }
+
+  private object ResolutionResult {
+
+    /**
+      * Indicates that a constraint was eliminated.
+      */
+    val elimination: ResolutionResult = ResolutionResult(Substitution.empty, Nil, progress = true)
+
+    /**
+      * Indicates that a constraint was resolved to a substitution.
+      */
+    def newSubst(subst: Substitution): ResolutionResult = ResolutionResult(subst, Nil, progress = true)
+
+    /**
+      * Indicates that a constraint was resolved to a set of constraints.
+      *
+      * This may be the same list of constraints as was given to the solver.
+      * In this case, `progress` is `false`.
+      */
+    def constraints(constrs: List[TypingConstraint], progress: Boolean): ResolutionResult = ResolutionResult(Substitution.empty, constrs, progress)
+  }
+
+  /**
+    * Converts the given unification error into a type error.
+    *
+    * Emulating logic from master branch
+    *
+    * ExpectType
+    * - pretend it's just unifyType
+    * - if mismatchedtypes then map the error to PossibleChecked or UnexpectedType
+    * - else return as is
+    *
+    * ExpectEffect
+    * - pretend it's just unifyType
+    * - if mismatchedEffects then map the error to possiblechecke or unexpectedeffect
+    * - else return as is
+    *
+    * ExpectTypeArguments
+    * - pretend it's just unifytype
+    * - if mismatchedbools or mismatchedarroweffects or mismatchedtypes then map the error to unexpectedarg
+    * - else return as is
+    *
+    * Match
+    * - mismatched types
+    *   - check for over/under applied
+    *   - else return as is
+    *     - mismatched bools -> mismatched bools
+    *     - mismatched effects
+    *   - check for mismatched arrow effects
+    *   - else return as is
+    *     - mismatched case sets -> mismatched case sets
+    *     - mismatched arity -> mismatched arity
+    *     - rigid var -> mismatched types
+    *     - occurs check -> occurs check
+    *     - undefined label -> undefined label
+    *     - non-record type -> non-record type
+    *     - undefined predicate -> undefined predicate
+    *     - non-schema type -> non-schema type
+    *     - no matching instance
+    *   - check for specific instance
+    *     - toString
+    *     - eq
+    *     - ord
+    *     - hash
+    *     - ?
+    *       - (other cases should be impossible on this branch)
+    */
+  // TODO ASSOC-TYPES This translation does not work well
+  // TODO ASSOC-TYPES because provenance is not propogated properly.
```

Comment:
I am OK with this for now, but whats the overall plan for this?

Reply:
Like whats the source of the messy nice? Is it that provenance is not enough? Or the constraint do not carry enough information? (which would be provenance, I guess)?

Reply:
I don't think that provenance is good enough because there are many constraints that contribute to an error, but we consider whichever one we actually fail on to be the cause.

Reply:
I see. But it seems hard to do better with a lot more stuff.

Reply:
The current errors seem okayish